### ISSUE TITLE
Pass set of effects + props as argument, instead of raw element

### ISSUE THREAD

**Issue by dy** [spectjs/spect#86]
React, inferno, haunted and alike, provide super-useful alternative: all it takes for component is just taking props and returning html:

```js
const App = ({name}) => html`...`
```

Losing that simplicity is a bummer.
With spect that is
```js
const App = el => $(el).html`...`
```

Besides, things like 
```js
const App = el => $(el).html`${ data.map(item => $`<${ Card } ...${item}/>`) }`
```
are mess:
- redundant wrapping
- no use in having `el` as props, since we ought to wrap it
- `html` is mixed with `$`
- `Card` component is inserted.

Compare haunted:
```js
const App = ({data}) => data.map(item => html`<app-card ...${item}/>`)
```

**Comment by dy** [comment]
That at core conflicts with aspects.

(also after playing around - haunted seems to break every here and then, also it needs normal extension of htm, also I hard to imagine easy integration with anything like eg. [anima](https://ghub.io/anima)).

If we assign multiple pieces of functionality, we have to have something tangible - element or wrapper, or, __planned rendering queue__.

---

We can try to mitigate these drawbacks:
```js
const App = ({ data, html }) => data.map(Card)
```
But then we have to pass wrapped elements, not the real ones, or expose special attributes on elements props + effects. From #48, #63: main argument to return direct element was to decouple `$` and `use`, which is practically hard.
Passing direct element creates redundant manual wrapping, messes up clean effects, messes up passing props.

Actually, passing effects along with props into some wrapping thing would make it consistent with html renderer `<div html=${rawHTML}/>` - this way props provided via `html` effect are clear indicator of being passed into wrapper.

### To recap.
Merging props with effects and passing wrapped target to aspect allows:
+ get rid of redundant wrapping `(el) => $(el).html...` → `({html}) => html... `
  + that gets rid of `$` as constructor
  + that is compatible with tape-like wrapped effects
  + that makes more compact use of effects
+ that resolves #84 
+ that makes effects a better indicator of planned changes
+ that shrinks unelegant implementation of nested items:

```js
const App = el => $(el).html`${ data.map(item => $`<${ Card }/>`) }`
// turns into
const App = ({ html }) => html`${ data.map(Card) }`
```

**Comment by dy** [comment]
The specifics of proxies.
```js
html`<a href=/ ></a>`
$els.href // returns href of first item?

html`<${Card} items=${data}/>`
$els.items // returns prop of the first item?
```

1. So it seems that wrapped sets only provide effects, but if we pass them to components via
```js
html`<${Card} items=${data}/>`
function Card($el) {
$el.html
$el.items
}
```
There props are on the same level as effects.
This turns effects namespace into easily extendable by user data.

? what if we'd just hide props behind an effect?
```js
({ prop, html }) => html`${ prop('data').map(() => html`<${Card} ...${item}/>`) }`
// that is explicit, but unclear why `html` is passed, but custom props `items` aren't.
```

2. Another question - if element has `<div items=${data} html=${'...content'} use=${aspect} />` - why `html` effect is handled instantly, but `items` is passed to renderer? Why `html` is passed to renderer, but not as value, but as external effect?

So: 1., 2. (conceptually), fx/attrs html conflict, subhtml conflict.

? What if we pass props as extension of effects via `Object.create`?

**Comment by dy** [comment]
? What if we pass props as extension of effects via `Object.create`?
? Or vice-versa, we extend element via `Object.create` or `Promise`, so that effects are available on the element?

So,
1. props are passed as is `({ items, ...props }) => [...items]`
2. effects are defined in prototype `({ items, html, ...props }) => html(...)` - (looks similar to redux methods), but can be overridden by props
3. effects are bound to target/collection, but target/collection should not have props exposed, ie. `$el => ...` - is not actually just `$el`, but `Object.create($el)` with assigned props (to avoid overriding props via proxy/etc).
4. In construct like `({ items, href, etc, ...$el }) => ...` - `$el` is not a proxy anymore, but destructured - props, (not data in prototype). So `$el` doesn't represent collection, if destructured. Wrapped elements are not destructured too `({ ...rest}) => rest[0] //undefined`, which is good. 
5. Single element can be destructured as `([el]) => {...}`.
6. Should we double-extend - a. `props` <- `$element` <- `element`? b. `$element` <- `element`? c. `props` <- `$element`?
  - b. rest spread doesn't destructure props, but only effects `({ items, ...$el })` - including elements in collection. It requires `$el` to be not promise, but simple object.
  - b. case we can't easily debug passed props, they're not apparent.
  - a. `({...props}) => props.items // no props.html`. in b. `({ ...fx }) => fx.html // no fx.items`. See table

 | `props` <- `$element` <- `element` | `$element` <- `element` | `props` <- `$element`
---|---|---|---
debugging | props are visible | props are hidden :x: | props are visible
`({items, html, ...props}) => {}` | `props` are rest props | `props` are `fx`, but not `$el` | `props` are rest props 
`([el, el2, ...rest]) => {}` | `props` are not array, although could be | el is taken as first element, but el2 can be taken from prototype `form` - mega-confusing :x: | ok |
`$el => $el.href` | leaks/mirros real DOM | leaks/mirrors DOM | undefined prop/effect
`el => el.href` | `el` is not real element | `el` is not real element | `el` is not element at all
`$els.use($el => {})` | `el`, wrapped with effects | `el`, wrapped with effects | ~ pure `$el`, needs destructuring via `[el]`

**Comment by dy** [comment]
Same time, all the 3 hide the fact in debugger that that's actually `$` passed into `use`.
The nicest look gives plain array - it shows both props and wrapped elements, as well as has methods in prototype.
? What if we'd just keep array, but extend id with passed props (not effects)?

**Comment by dy** [comment]
Ok, Arrays have pros.
Should we combine use calls into a single wrapped-set call?
This saves redundant calc time.
But effect reducers are called on per-element basis. Elements may have overlapping states/props/etc.
Since an aspect can match multiple elements with various props, they must be called per-element.

**Comment by dy** [comment]
Fixed in https://github.com/spectjs/spect/commit/ce3909e010c174f9ef5e71d2a9729cdfa04e161f as arrays.
